### ISSUE TITLE
Horizontal `stack` does not correctly resize its children

### ISSUE THREAD

**Issue by tingerrr** [typst/typst#2422]
### Description

A horizontal `stack` will overflow out of its enclosing container if it cannot fit its contents instead of resizing them.

```typst
#set page(width: 4cm)
#stack(dir: ltr, [Hello], 5pt, lorem(10))
```

![image](https://github.com/typst/typst/assets/137803093/ea6fee85-a0bd-4bda-8eec-4f8bcee4af4a)

See: https://github.com/typst/typst/issues/2410#issuecomment-1766371491
            

### Operating system

Web app

### Typst version

- [X] I am using the latest version of Typst

**Comment by Enivex** [comment]
I believe the issue is that the last column has 100% of the width available, instead of subtracting the space already used for the previous entries.

**Comment by EpicEricEE** [comment]
I just looked into this a bit and it's kind of a dilemma. For example (taken from the `visualize/shape-aspect` test)
```typ
#block(width: 100pt, stack(
  dir: ltr,
  square(width: 20pt),
  circle(width: 20%),
))
```
currently makes the square and circle have the same size, which makes sense: the 20% width is resolved to the total size of the available parent space (100pt). If we now were to subtract the already used up width before laying out the circle (which would fix this issue), it would only have 80pt available space, so its width would shrink to `20% * 80pt = 16pt`.

**Comment by psimons2** [comment]
Not really, it should only affect fractional spacing and other boxes that can shrink. 20% of 100pt should still be 20pt.

**Comment by laurmaedje** [comment]
@EpicEricEE for the Y axis, there exist both `full` and `size.y` in `Regions` for this reason. `size.y` is what's left and `full` is used for relative sizing. I guess we'd need a similar mechanism for horizontal. I believe it actually existed at some point and I removed it, but I can't remember the reason anymore. might've just been because it was unused (i.e. always the same as `size.x`) at the time.

**Comment by Andrew15-5** [comment]
> I believe it actually existed at some point and I removed it, but I can't remember the reason anymore. might've just been because it was unused (i.e. always the same as `size.x`) at the time.

Does this mean that it's easily fixable? I was about to propose a new stack field since it doesn't feel solvable to me.

**Comment by laurmaedje** [comment]
I am not certain, but I think it's fixable without a new field.

**Comment by laurmaedje** [comment]
@LinusU For now, I would recommend to use `grid` instead

**Comment by LinusU** [comment]
@laurmaedje glorious! Thank you for the pointer!

This code:

```typst
#set page(width: 512 * 0.75pt, height: auto, margin: 2pt)
#set text(black, font: "Roboto")

#let data = ("createdAt":datetime(year:2025,month:5,day:15,hour:16,minute:7,second:10),"printerName":"Custom 4 (för leveransbong)")


#grid(
  columns: (auto, auto, auto),
  gutter: 8pt,
  text(size: 18pt, data.createdAt.display("[year]-[month]-[day]")),
  align(center, text(size: 18pt, data.printerName)),
  text(size: 18pt, data.createdAt.display("[hour]:[minute]:[second]")),
)


#v(64pt)
```

Yields this result:

![Image](https://github.com/user-attachments/assets/16ead4eb-1d8d-4dd0-895c-922c44552815)

**Comment by LinusU** [comment]
I've found another example (that I also managed to fix by using `grid` instead).

Note that I only found this on 1 in 50 examples, so it only seems to be a specific width that triggers it. Quite possibly that the second text here extends 63pt off screen...

```typst
#set page(width: 512 * 0.75pt, height: auto)

#set page(margin: 2pt)
#set text(black, font: "Roboto")

#v(64pt)

#set text(size: 24pt)

#stack(
  dir: ltr,
  box(width: 63pt, text("2 x")),
  text("Charles de Fère Blanc de Blancs Brut Frankrike glas"),
)

#v(64pt)

#grid(
  columns: (63pt, 1fr),
  text("2 x"),
  text("Charles de Fère Blanc de Blancs Brut Frankrike glas"),
)

#v(64pt)
```

![Image](https://github.com/user-attachments/assets/87a1f293-67b5-472a-992f-d832b59d04ab)
